sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,98 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef np.ndarray parent_arr
+    cdef np.ndarray size_arr
+    cdef ITYPE_t next_label
+    cdef ITYPE_t *parent
+    cdef ITYPE_t *size
+
+    def __init__(self, N):
+        self.parent_arr = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size_arr = np.hstack((np.ones(N, dtype=ITYPE),
+                                   np.zeros(N-1, dtype=ITYPE)))
+        self.parent = (<ITYPE_t *> self.parent_arr.data)
+        self.size = (<ITYPE_t *> self.size_arr.data)
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.parent[m] = self.next_label
```

Comment:
I'm a bit surprised by the use of parent here and parent_arr below. Why not just use a memoryview to get speed and assurances?

Reply:
I believe the answer is that I was stealing old code from back when I was even worse at cython than I am now.